File: zenoh_pc_nodes/cpp/src/mpu6050_sub.cpp
#include <zenoh_ros/ZenohRosPC.h>
#include <zenoh_ros/sensor_msgs/z_Imu.h>
#include <iostream>
#include <iomanip>

class MPU6050SubscriberNode : public ZenohNode {
public:
    MPU6050SubscriberNode() : ZenohNode("mpu6050_subscriber") {
        std::cout << "[Node] " << this->z_get_name() << " has been started\n";

        // Create subscription to topic "robot/mpu6050" with standard ROS 2 z_Imu interface
        sub_ = this->z_create_subscription<z_Imu>(
            "robot/mpu6050",
            [this](const z_Imu& msg) -> void {
                this->listener_callback(msg);
            },
            10
        );
    }

private:
    ZenohSubscription<z_Imu>* sub_ = nullptr;

    void listener_callback(const z_Imu& msg) {
        std::cout << std::fixed << std::setprecision(2)
                  << "[IMU RECV] [stamp: " << msg.header.stamp.sec << "." << msg.header.stamp.nanosec << "] Accel: (" 
                  << std::setw(6) << msg.linear_acceleration.x << ", " 
                  << std::setw(6) << msg.linear_acceleration.y << ", " 
                  << std::setw(6) << msg.linear_acceleration.z << ") m/s² | Gyro: ("
                  << std::setw(6) << msg.angular_velocity.x << ", " 
                  << std::setw(6) << msg.angular_velocity.y << ", " 
                  << std::setw(6) << msg.angular_velocity.z << ") rad/s\n";
    }
};

int main(int argc, char** argv) {
    (void)argc;
    (void)argv;

    // Connect to ESP32-S3 SoftAP IP (192.168.4.1)
    ZenohConfig config;
    config.host = "192.168.4.1";
    config.port = 7447;

    if (!ZenohNode::init(config)) {
        std::cerr << "Failed to initialize Zenoh session!\n";
        return -1;
    }

    {
        MPU6050SubscriberNode node;
        node.z_spin();
    }

    return 0;
}

File: zenoh_pc_nodes/python/mpu6050_sub.py
#!/usr/bin/env python3
import sys
import os

# Ensure the shared zenoh_ros package can be imported
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '../../shared_libraries/python')))

from zenoh_ros import ZenohNode, ZenohConfig
from zenoh_ros.sensor_msgs import z_Imu

class MPU6050SubscriberNode(ZenohNode):
    def __init__(self) -> None:
        super().__init__("mpu6050_subscriber")
        print(f"[Node] {self.z_get_name()} has been started")

        # Create subscription to topic "robot/mpu6050" with standard ROS 2 z_Imu type
        self.sub = self.z_create_subscription(
            z_Imu,
            "robot/mpu6050",
            self.listener_callback,
            10
        )

    def listener_callback(self, msg: z_Imu) -> None:
        print(
            f"[IMU RECV] [stamp: {msg.header.stamp.sec}.{msg.header.stamp.nanosec:09d}] "
            f"Accel: ({msg.linear_acceleration.x:6.2f}, {msg.linear_acceleration.y:6.2f}, {msg.linear_acceleration.z:6.2f}) m/s² | "
            f"Gyro: ({msg.angular_velocity.x:6.2f}, {msg.angular_velocity.y:6.2f}, {msg.angular_velocity.z:6.2f}) rad/s"
        )


def main() -> None:
    # Connect to ESP32-S3 SoftAP IP (192.168.4.1)
    ZenohNode.init(host="192.168.4.1", port=7447)
    node_instance = None
    try:
        node_instance = MPU6050SubscriberNode()
        node_instance.z_spin()

    except KeyboardInterrupt:
        print("\n[CTRL+C] Interrupted by user.")

    except Exception as e:
        print(f"Critical Error: {e}")

    finally:
        if node_instance is not None:
            print("Destroying the Zenoh Node...")
            node_instance.z_destroy()


if __name__ == '__main__':
    main()
